Isaac/client integration

ittullos opened the pull request:
WIP Client Integration

- Align requests and specs with the server's expected data structure to match client requests.
- Add a new rake task to seed the database with residents and journeys
- Introduce a new ENV variable to pass the Cognito audience to the `verify!` method.

Still getting some errors having to do with passing the resident list
pdf to the server as a param and passing the cognito audience value
to the verify! method.  I'll include the error messages in the comments
of the relevant request.

I do still have some failing tests here as well that need to be fixed.

ittullos committed:
Seed journeys

This commit adds a seed file for journeys. This is to be used in
development to test the client locally. The Rakefile has also
been updated to include a task to seed the database.

ittullos commented on app/app.rb:
```diff
@@ -28,34 +28,32 @@
 post '/user/residents' do
-  unless params[:file] && params[:file]['tempfile']
+  unless parsed_params[:file] && parsed_params[:file]['tempfile']
     status 400
```
Running into issues here because the parsed_params structure does not contain a 'tempfile' attribute. When I change line 56 to `unless parsed_params.fetch("file")` I get a 500 error. Here is the structure I am getting back from parsed_params:

`{"file"=>
  {"canceled"=>false,
   "assets"=>
    [{"uri"=>
       "file:///Users/rafiki/Library/Developer/CoreSimulator/Devices/301E4715-A44D-4406-8873-B25EA2E93581/data/Containers/Data/Application/37D9B03D-4615-4740-9EC0-B95E2B4F8413/Library/Caches/ExponentExperienceData/%2540anonymous%252Ffrontend-53f2857d-bbf9-4422-b861-b0d116a53d29/DocumentPicker/E953D495-6DE6-4251-ABD7-13238ADDEDFB.pdf",
      "name"=>"neighborhoodlist-Isaac+Tullos (1).pdf",
      "size"=>56918,
      "mimeType"=>"application/pdf"}]}}`

zbarnard1 replied:
Google suggests we should be using FormData as the CONTENT_TYPE rather than JSON, so let's try something like this:
```
import axios from 'axios';
import { Platform } from 'react-native';
import * as DocumentPicker from 'expo-document-picker';

async function uploadFile() {
  const result = await DocumentPicker.getDocumentAsync({ type: 'application/pdf' });

  if (result.type === 'success') {
    const formData = new FormData();
    formData.append('file', {
      uri: result.uri,
      name: result.name,
      type: result.mimeType,
    });

    try {
      const response = await axios.post('http://your-server-url/user/residents', formData, {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      });
      console.log(response.data);
    } catch (error) {
      console.error(error);
    }
  }
}
```

I expect the Sinatra code here to still work, since Sinatra should be saving that file to a temporary location (thus the `tempfile` thing). Give that a go and let me know if that solves it. It might be helpful to see the ReactNative code as well.

ittullos commented on app/lib/json_web_token.rb:
```diff
@@ -18,7 +22,7 @@ def verify!
         verify_iss: true,
-        aud: 'P4L-API',
+        aud: ENV['COGNITO_APP_CLIENT_ID'],
         verify_aud: true,
```
I'm using an environment variable here to store the client_id. I can see the correct value coming through in the debug logs above but still getting a 401 error. I'm getting the value of the client_id from the AWS console.

zbarnard1 replied:
I suspect that the access token doesn't have the `aud` claim. The access token will have this value as it's `client_id`, but it doesn't seem to have an `aud`. https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-the-access-token.html

The easiest fix is probably just to NOT verify the `aud` at all. I think we can remove both the `aud` and `verify_aud` keyword args from this `decode` method call and it won't try to verify it for either token. There aren't any other apps in the ecosystem so there's no risk that an attacker will use a token from another app to gain access to this one.
Alternatively, we split the `verify!` method into 2 versions, one for the ID token which DOES verify the `aud` claim, and one for the access token which does not. This is a little more work and more code to maintain, but gives us slightly more security. Your call, I'm OK with either.

zbarnard1 replied:
A third option is to not verify the audience, but manually check the `client_id` against the COGNITO_APP_CLIENT_ID. We get the payload back from `verify!`, so we can do that in the authentication middleware. Maybe for now we just don't verify the audience and we can add this later if there's another API added to the mix.

zbarnard1 commented on spec/tasks/seed_residents_spec.rb:
```diff
@@ -0,0 +1,33 @@
+
+RSpec.describe 'Seeding the database with residents', :task do
+  before do
+    Resident.delete_all # Clear the database before each test
```
You shouldn't generally need to do this, DatabaseCleaner does this already. Let me know if this isn't the case.

ittullos committed:
Seed residents

This commit adds a seed file for residents. It also updates the Rakefile
to include a new task for seeding residents.

ittullos committed:
Remove aud checking in jwt verify! method

ittullos committed:
Change route return values and tests

Changes the return values of the route endpoints to be more consistent.
The tests have been updated to reflect the new return values. Also
created a new helper method called parsed_params to help with parsing
the params passed in the requests.

ittullos commented:
Got everything fixed and cleaned up. Let me know what you think! 🙌

zbarnard1 reviewed (approved):
Looks really good, ship it!
